### DISCUSSION TITLE
state of py > exe

### DISCUSSION THREAD

**Discussion by dtoft** [py2exe/py2exe#197] [General]
What's happened with the state of py > exe tech? It's been a while (2017) since I've owned this process and, coming back, it seems like py2exe is no longer the leading open source package, at least in terms of support community. I used py2exe for a while with great results. But now it seems to have fallen behind python + package updates (doesn't work for my current need)

Now it seems like everyone is using PyInstaller, which is supposed to be better for the latest python. But after trying it for a few hours, I'm quickly disappointed. It doesn't build a professional distribution package like py2exe and seems to be a quick-fix for deploy-python-to-users-without-python (vs a more robust desktop app distribution). It doesn't handle app icons and Windows Properties Details well

So now I'm wondering if I should invest more time into getting py2exe to work and deal with what I expect to be ongoing issues compiling the data analysis packages. But it's got me wondering, what happened? What am I missing? Why did the community abandon py2exe? Why isn't PyInstaller building for professional deployment? Is there a different package I should be looking at?

Wondering what the community thinks!

**Comment by albertosottile** [comment]
Hello, thanks for this input. Perhaps we can have a fruitful discussion here.

> Now it seems like everyone is using PyInstaller, which is supposed to be better for the latest python. But after trying it for a few hours, I'm quickly disappointed.

This is the reason why I have been maintaining py2exe for the past 4 years...

> So now I'm wondering if I should invest more time into getting py2exe to work and deal with what I expect to be ongoing issues compiling the data analysis packages. 

This is entirely up to you, of course any contribution would be welcome. Just be aware that there are [some fundamental issues](https://github.com/py2exe/py2exe/issues/191#issuecomment-1774211437) that will most likely prevent to use `bundle_files <= 2` with any Python version newer than 3.11

> What happened? What am I missing? Why did the community abandon py2exe?

I do not have definitive answers to this, I can only provide my point of view. IMHO, this boils down to three main reasons:
- freezing executables in Python has always been and still is a second-class citizen. No active upstream effort towards this goal has ever happened (actually, some changes were probably for the worse...)
- developing tools like py2exe in general, and py2exe specifically, is relatively hard, time-consuming, error-prone, bound-to-be-broken, almost untestable... not really some encouraging points when thinking about contributing to an open source project
- PyInstaller has the biggest community support because it is multi platform. That being said, also PyInstaller is struggling in finding contributors, maintainers, and support in general

> Why isn't PyInstaller building for professional deployment?

This you should ask to the PyInstaller guys. I remember that some requests (e.g. support one shared library bundle across two executables) were ignored for years due to the lack of development resources. 

> I used py2exe for a while with great results. But now it seems to have fallen behind python + package updates (doesn't work for my current need)

I would just like to conclude with this: I do not believe that any unofficial freezing solution could ever be compatible with ALL the Python ecosystem, unless an effort is made upstream to standardize the whole process and the requirements a package must have to be successfully frozen. You can have a look at the py2exe or PyInstaller hooks to have a glimpse on the shenanigans that developers constantly add in their packages (and this is just what we know, perhaps a tenth of PyPI?).

**Reply by dtoft** [reply]
Thanks for the response,

> I have been maintaining py2exe for the past 4 years...

God bless you. I can't imagine the headache of trying to make it work for any number of packages and versions. Python's blessing and its curse. I can appreciate the amount of effort required to keep up

> I do not believe that any unofficial freezing solution could ever be compatible with ALL the Python ecosystem

Agree. But the bug I found was in the matplotlib hook, which I expect is a major. I read through some of the recent changes related to it and it sounded a bit improvised, which is when I got bummed. Maybe my issue is bundle_files related but this came after I upgraded python from reading setup.py was deprecated for freeze, which I was OK with, time to upgrade anyway. But then features are breaking for updated python too so...

I can understand the multi-platform feature attracting more support. Looks like a lot of macos use over there. Have you thought about merging the best of py2exe with PyInstaller or are the core approaches too different?

At a higher level, I guess I worry if python app distribution isn't easy-ish, end-to-end python distribution will always be between (fade to?) the C family toolkit (Discord uses [Squirrel](https://github.com/Squirrel/Squirrel.Windows), which sounds great) and the web toolkit (even Jupyter Lab Desktop uses Electron). I would expect orgs like Qt, Anaconda to care about that. Or do most feel the future is distributing python more on web eg Anaconda doing [pyscript](https://pyscript.net)?

**Reply by albertosottile** [reply]
> Agree. But the bug I found was in the matplotlib hook, which I expect is a major.

`matplotlib` is part of the CI in this repository and so it is "guaranteed" to be functional, but only with `bundle_files == 3`. At some point, I had to stop active support for third-party packages and `bundle_files <= 2`, because I simply do not have the capacity to provide that. This could always change if this project gained 2-3 more maintainers but, for the reasons discussed before, this seems every day more unlikely.

**Reply by dtoft** [reply]
3 more maintainers just to stay on top of versioning + more to push the package forward. Where would you want to see it go if there were 6 more contributors?
